lib/Dialect/TritonGPU/Transforms/WarpSpecialization/PartitionLoops.cpp

```diff
@@ -16,103 +19,337 @@ using namespace mlir;
+    auto step = b.mapping.lookupOrDefault(forOp.getStep());
+    SmallVector<Value> initArgs;
+    for (auto idx : newLoopIndices) {
+      initArgs.push_back(forOp.getInitArgs()[idx]);
     }
-    // Store the result after the loop and load it after the base loop.
-    ImplicitLocOpBuilder b(result.getLoc(), baseLoop);
-    Value alloc = b.create<LocalAllocOp>(MemDescType::get(
-        ty.getShape(), ty.getElementType(), getSharedEncoding(ty),
-        SharedMemorySpaceAttr::get(ty.getContext()), /*mutable=*/true));
-    b.setInsertionPointAfter(loop);
-    b.create<LocalStoreOp>(result, alloc);
-    b.setInsertionPointAfter(baseLoop);
-    Value output = b.create<LocalLoadOp>(ty, alloc);
-    b.create<LocalDeallocOp>(alloc);
-    baseLoop.getResult(resultIdx).replaceAllUsesWith(output);
+    auto newForOp =
+        b.builder.create<scf::ForOp>(forOp.getLoc(), lb, ub, step, initArgs);
+    newForOp->setAttrs(forOp->getAttrs());
+    newForOps.push_back(newForOp);
+
+    b.mapping.map(forOp.getInductionVar(), newForOp.getInductionVar());
+
+    auto oldIterArgs = forOp.getRegionIterArgs();
+    auto newIterArgs = newForOp.getRegionIterArgs();
+    for (auto [newIdx, oldIdx] : llvm::enumerate(newLoopIndices)) {
+      b.mapping.map(oldIterArgs[oldIdx], newIterArgs[newIdx]);
+      b.mapping.map(forOp.getResult(oldIdx), newForOp.getResult(newIdx));
+    }
+
+    b.builder.setInsertionPointToStart(newForOp.getBody());
   }
 
-  // Delete everything else.
-  eraseOtherPartitions(loop, schedule, partition);
+  cloneOpsInBlock(forOp.getBody(), builders, schedule);
 
-  return std::move(block);
+  for (auto newForOp : newForOps) {
+    WarpSchedule::eraseFrom(newForOp);
+  }
 }
 
-//===----------------------------------------------------------------------===//
-// partitionLoop
-//===----------------------------------------------------------------------===//
+void cloneIfOp(scf::IfOp ifOp, SmallVector<WgBuilder> &builders,
+               const WarpSchedule &schedule) {
+  auto partition = getPartition(ifOp, schedule);
+  assert(partition);
+  SmallVector<size_t> partitionIndices;
+
+  if (partition == schedule.getRootPartition()) {
+    for (size_t i = 0; i < builders.size(); ++i) {
+      partitionIndices.push_back(i);
+    }
+  } else {
+    partitionIndices.push_back(partition->getIndex());
+  }
+
+  SmallVector<scf::IfOp> newIfOps;
+  for (size_t idx : partitionIndices) {
+    auto &b = builders[idx];
+    SmallVector<Type> newIfResultTypes;
+    for (auto [idx, result] : llvm::enumerate(ifOp.getResults())) {
+      newIfResultTypes.push_back(result.getType());
+    }
+    auto cond = b.mapping.lookupOrDefault(ifOp.getCondition());
+    auto newIfOp = b.builder.create<scf::IfOp>(
+        ifOp.getLoc(), newIfResultTypes, cond, ifOp.elseBlock() ? true : false);
+    newIfOp->setAttrs(ifOp->getAttrs());
+    newIfOps.push_back(newIfOp);
+
+    // map arguments and results
+    for (auto [oldRes, newRes] :
```

Comment:
can you add a helper `mapRange` that maps a `ValueRange` elementwise to a `ValueRange`?